src/collaborate.rs

```diff
@@ -150,33 +245,112 @@ impl Collaboration {
                             self.my_key.clone(),
                         )?;
                         self.aggregated_key = Some(aggregated.clone());
-
                         self.state = true;
                     }
                 } else {
-                    let keys: ParticipantKeys = parse_keys(data)
+                    let keys: ParticipantKeys = parse_keys(data.clone())
                         .map_err(|_| BitVMXError::InvalidMessageFormat)?
                         .first()
                         .unwrap()
                         .1
                         .clone();
 
-                    keys.mapping.iter().for_each(|(pubkey_hash, key)| {
-                        let pubkey_hash: PubKeyHash = pubkey_hash
+                    //validates leader's verification key
+                    let leader_verification_key = self
+                        .participant_verification_keys
+                        .get(&self.leader.pubkey_hash)
+                        .unwrap();
+
+                    let message = crate::comms_helper::construct_message(
+                        &self.collaboration_id.to_string(),
+                        &data,
+                        timestamp,
+                    )?;
+
+                    let verified = program_context.key_chain.verify_rsa_signature(
```

Comment:
It seems that we are not enforcing the result of the verification, only logging it. If verified is false for the leader's key the execution flow continues, accepting the leader's key as valid.